tidy3d/plugins/autograd/utilities.py

```diff
@@ -158,3 +162,69 @@ def chained(array: NDArray):
         return reduce(lambda x, y: y(x), funcs, array)
 
     return chained
+
+
+def scalar_objective(func: Callable = None, *, has_aux: bool = False) -> Callable:
+    """Decorator to modify a function to return the `.data` attribute of an xarray.DataArray.
+
+    If the function returns an xarray.DataArray, this decorator extracts the `.data` attribute.
+    Additionally, it checks if the return value is scalar and real, raising an exception if it is not.
+    The result is also squeezed to remove single-dimensional entries.
+
+    Parameters
+    ----------
+    func : Callable
+        The function to be decorated.
+    has_aux : bool = False
+        Whether the function returns auxiliary data.
+
+    Returns
+    -------
+    Callable
+        The decorated function.
+
+    Raises
+    ------
+    Tidy3dError
+        If the return value is not scalar or not real.
+    """
+
+    if func is None:
```

Comment:
what is this for? if you want to do
 ```py
f = scalar_objective(has_aux=True)

@f
def objective(....)
```
?

Reply:
yeah, although I look at it mostly through the decorator syntax:
```python
@scalar_objective  # implicit has_aux=False
def objective(x):
    ...
```
vs
```python
@scalar_objective(has_aux=True)
def objective(x):
    ...
```